wlauto/tests/test_environment.py

```diff
@@ -0,0 +1,289 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+# pylint: disable=W0231,W0613,E0611,W0603,R0201
+from unittest import TestCase
+
+from nose.tools import assert_equal, assert_raises
+
+from wlauto.utils.environment_manager import (init_environment, reset_environment,
+                                              activate_environment, once,
+                                              once_per_class, once_per_instance)
+
+class TestClass(object):
+
+    def __init__(self):
+        self.count = 0
+
+    @once
+    def initilise_once(self):
+        self.count += 1
+
+    @once_per_class
+    def initilise_once_per_class(self):
+        self.count += 1
+
+    @once_per_instance
+    def initilise_once_per_instance(self):
+        self.count += 1
+
+
+class SubClass(TestClass):
+
+    def __init__(self):
+        super(SubClass, self).__init__()
+
+    @once
+    def initilise_once(self):
+        self.count += 1
+
+    @once_per_class
+    def initilise_once_per_class(self):
+        self.count += 1
+
+    @once_per_instance
+    def initilise_once_per_instance(self):
+        self.count += 1
+
+
+class SubSubClass(SubClass):
+
+    def __init__(self):
+        super(SubSubClass, self).__init__()
+
+    @once
+    def initilise_once(self):
+        self.count += 1
+
+    @once_per_class
+    def initilise_once_per_class(self):
+        self.count += 1
+
+    @once_per_instance
+    def initilise_once_per_instance(self):
+        self.count += 1
+
+
+class AnotherClass(object):
+
+    def __init__(self):
+        self.count = 0
+
+    @once
+    def initilise_once(self):
+        self.count += 1
+
+    @once_per_class
+    def initilise_once_per_class(self):
+        self.count += 1
+
+    @once_per_instance
+    def initilise_once_per_instance(self):
+        self.count += 1
+
+
+class EnvironmentManagementTest(TestCase):
+
+    def test_duplicate_environment(self):
+        init_environment('ENVIRONMENT')
+        assert_raises(ValueError, init_environment, 'ENVIRONMENT')
+
+    def test_reset_missing_environment(self):
+        assert_raises(ValueError, reset_environment, 'MISSING')
+
+    def test_reset_current_environment(self):
+        activate_environment('CURRENT_ENVIRONMENT')
+        t1 = TestClass()
+        t1.initilise_once()
+        assert_equal(t1.count, 1)
+
+        reset_environment()
+        t1.initilise_once()
+        assert_equal(t1.count, 2)
+
+    def test_switch_environment(self):
+        activate_environment('ENVIRONMENT1')
+        t1 = TestClass()
+        t1.initilise_once()
+        assert_equal(t1.count, 1)
+
+        activate_environment('ENVIRONMENT2 ')
+        t1.initilise_once()
+        assert_equal(t1.count, 2)
```

Comment:
To complete this test, I would re-activate ENVIRONMENT1, call `intialise_once()` again and make sure that `t1.count` does not increment